lib/puppet/functions/ldapquery/query.rb

```diff
@@ -16,143 +16,276 @@
+  #   # ---
+  #   # base: dc=acme,dc=example,dc=com
+  #   # host: ldap.example.com
+  #   # auth:
+  #   #   method: simple
+  #   #   username: ldapuser
+  #   #   password: ldappassword
+  #
+  #   ldapquery::query(
+  #     undef, # `base` can also be set in `ldapquery.yaml`
+  #     '(objectClass=dnsDomain)',
+  #     ['dc'],
+  #   )
+  # @example Querying an LDAP server using TLS and the system cert store
+  #   ldapquery::query(
+  #     'dc=acme,dc=example,dc=com',
+  #     '(objectClass=dnsDomain)',
+  #     ['dc'],
+  #     {
+  #       host       => 'ldap.example.com',
+  #       port       => 636,
+  #       encryption => {
+  #         method => simple_tls,
+  #       },
+  #       auth       => {
+  #         method   => simple,
+  #         username => 'ldapuser',
+  #         password => 'ldappassword',
+  #       },
+  #     },
+  #   )
+  # @example Querying an LDAP server using TLS with a custom CA certificate
+  #   ldapquery::query(
+  #     'dc=acme,dc=example,dc=com',
+  #     '(objectClass=dnsDomain)',
+  #     ['dc'],
+  #     {
+  #       host       => 'ldap.example.com',
+  #       port       => 636,
+  #       encryption => {
+  #         method      => simple_tls,
+  #         tls_options => { ca_file => '/path/to/custom-ca.crt' },
+  #       },
+  #       auth       => {
+  #         method   => simple,
+  #         username => 'ldapuser',
+  #         password => 'ldappassword',
+  #       },
+  #     },
+  #   )
+  # @example Querying an LDAP server on the standard port but then encrypting all traffic using STARTTLS
+  #   ldapquery::query(
+  #     'dc=acme,dc=example,dc=com',
+  #     '(objectClass=dnsDomain)',
+  #     ['dc'],
+  #     {
+  #       host       => 'ldap.example.com',
+  #       port       => 389, # Default included for clarity
+  #       encryption => {
+  #         method => start_tls,
+  #       },
+  #       auth       => {
+  #         method   => simple,
+  #         username => 'ldapuser',
+  #         password => 'ldappassword',
+  #       },
+  #     },
+  #   )
+  # @example Specifying a 1 minute `search_timeout`, (and `connect_timeout` via `ldapquery.yaml`)
+  #   # /etc/puppetlabs/puppet/ldapquery.yaml
+  #   # ---
+  #   # base: dc=acme,dc=example,dc=com
+  #   # host: ldap.example.com
+  #   # auth:
+  #   #   method: simple
+  #   #   username: ldapuser
+  #   #   password: ldappassword
+  #   # connect_timeout: 30
+  #
+  #   ldapquery::query(
+  #     undef, # Use `base` from `ldapquery.yaml`
+  #     '(objectClass=dnsDomain)',
+  #     ['dc'],
+  #     {}, # Use `ldapquery.yaml`
+  #     'sub',
+  #     60,
+  #   )
+  dispatch :query do
+    param 'Optional[String[1]]', :base
+    param 'Optional[String[1]]', :filter
+    param 'Array[String[1],1]', :attributes
+    optional_param 'LDAPArgs', :ldap_args
+    optional_param 'Enum["sub","base","single"]', :scope
+    optional_param 'Integer[1]', :search_timeout
+    return_type 'Array'
   end
```

Comment:
From my personal view i do not get why that loading of the new config file and extended value support must be hidden beneath a new dispatch. Also that optional base and filter are not optional_param is strange for me.

`ldapquery::query(undef, undef, ['uid'])` looks strange.

Reply:
I wanted to retain the current interface alongside the new one.  Some people might have calls to the function spread out across multiple manifests and forcing them to refactor them all at once isn't as friendly as publishing a version that works with both. The dispatches have to be unambiguous for this to work.
I'd imagine most people will only put their connection details in the config file and almost always also use a filter.  I made `undef` allowable for completeness.

Reply:
I think there needs to be a way of distinguishing what config file the code should use.  I really didn't want a function that tried to use both.  In the original options hash, only `server` was relevant to connecting to ldap, `base` and `scope` are really search arguments.  And do I merge `server` with `host` and `hosts`?  Having ldap_args being fed directly (after converting somethings to symbols as required), to the ruby ldap `search` method allows for any/all options to be specified without us trying to magically set anything based on a combination of other settings.

Reply:
Maybe it would be cleaner to just put the new code in `openldap::search` instead? (although I still think we just drop the original implementation at some point and not hack more stuff into it). And maybe it's not necessary to allow `undef` for `base` and `filter`.  Imposing that all calls have the first 3 arguments (base, filter, attributes) wouldn't be a hardship on users and might make your code look less magic and more readable.